### PR TITLE
Add Regression test to Shell32_apitest ShellExecuteEx to detect error…

### PR DESCRIPTION
**Pull request by Doug-Lyons**
… causing desk.cpl failure to appear after right-click on desktop and selecting Properties.

_Add additional regression test to shell32_apitest ShellExecuteEx._

JIRA issue: [CORE-16299](https://jira.reactos.org/browse/CORE-16299)

_No fixes are included here only a regression test to detect this error in the future._

### CODE REVIEW

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 1]
```diff
@@ -81,8 +155,67 @@ TestShellExecuteEx(const WCHAR* Name, BOOL ExpectedResult)
...
+    ShellExecInfo.lpFile = Name;
+    ShellExecInfo.lpParameters = Params;
+    ShellExecInfo.lpDirectory = NULL;
+    ShellExecInfo.lpVerb = L"open";
+
+    iResult = GetOSVendor(OSVendor, sizeof(OSVendor));
+    printf("OSVendor is '%s'.\n", OSVendor);
+
+    if(strcmp(OSVendor, "Microsoft") == 0)
+        skip("Test does not work on MS Windows.\n");
```
Comment (SergeGautherie) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: `skip()` without `return`?

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 2]
```diff
@@ -81,8 +155,67 @@ TestShellExecuteEx(const WCHAR* Name, BOOL ExpectedResult)
...
+
+    if(iResult == FALSE)
```
Comment (SergeGautherie) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: This should happen earlier.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 3]
```diff
@@ -81,8 +155,67 @@ TestShellExecuteEx(const WCHAR* Name, BOOL ExpectedResult)
...
+    Result = ShellExecuteExW(&ShellExecInfo);
+
+    ok(Result, "ShellExecuteEx lpFile %s failed. Error: %lu\n",
+                wine_dbgstr_w(Name), GetLastError());
+
+    if(strcmp(OSVendor, "Microsoft") == 0)
+        skip("Test does not work on MS Windows.\n");
+
+    if(iResult == FALSE)
+        skip("Unable to determine Operating System.\n");
```
Comment (SergeGautherie) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: These 2 checks duplicate previous code...

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 4]
```diff
@@ -6,8 +6,82 @@
...
+    CHAR file_name[25] = {"OperatingSystem.txt"};
+    CHAR cmdline[100] = {"ver > "};
+    FILE *fp;
+    CHAR OSVendor[100] = {0};
+    CHAR *myreturn;
+    INT pos;
+    char *space_ptr;
+    INT iResult;
+
+    if(VendorString && (length >= 20))
```
Comment (SergeGautherie) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: Missing space: `if ()`

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp (hunk 1 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: @SergeGautherie, Yes, I am using skip() without return's. For an existing example look at this:
modules\rostests\winetests\xinput1_3\xinput.c. Here there are five (5) skip() statements and no return statements. Others are readily found by just looking. I leave it as an exercise for the reader to determine why you want each of these different usage styles.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp (hunk 2 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: @SergeGautherie, I disagree. This statement and the previous one could occur is this order or reversed as you suggest and still result in the same reporting and outcome.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp (hunk 3 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: @SergeGautherie, This is true, but for a different type of test. The first messages apply to whether we could read the OS Vendor. The second set of messages apply to the results of ShellExecuteExW and why it did not work. Therefore, I contend that this does its job correctly.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp (hunk 4 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: Maybe true, but is it worth the effort to fix for a regressions test?

**Comment by Doug-Lyons** [comment]
I do not consider this a Work In Progress (WIP). I believe it to be complete and only needs reviews.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp (hunk 3 above)
Comment (SergeGautherie) [reply modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: As far as I can tell `iResult` and `OSVendor` have nothing to do with `ShellExecuteExW()`, so I see no purpose in checking/"reporting" the same things again.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp (hunk 2 above)
Comment (SergeGautherie) [reply modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: I usually do not check invalid data, but if that's what you want to do...

**Comment by SergeGautherie** [comment]
> complete

Really? 2 fopen(), 1 fclose(), many `return`: I love that, not!

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 5]
```diff
@@ -2,12 +2,87 @@
...
+
+    if (VendorString && (length >= 20))
+        strcpy(VendorString, "Unknown");
+    else
+        return FALSE;
+
+    fp = fopen(file_name, "r"); // read mode
+
+    if (fp != NULL)
+        return FALSE;
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: Close fp.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 6]
```diff
@@ -2,12 +2,87 @@
  * PROJECT:         ReactOS api tests
  * LICENSE:         GPLv2+ - See COPYING in the top level directory
  * PURPOSE:         Testing ShellExecuteEx
- * PROGRAMMER:      Yaroslav Veremenko <yaroslav@veremenko.info>
+ * PROGRAMMERS:     Yaroslav Veremenko <yaroslav@veremenko.info>
+ *                  Doug Lyons <douglyons@douglyons.com>
  */
 
 #include "shelltest.h"
+#include<stdio.h>
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: Add space before `<`.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 7]
```diff
@@ -2,12 +2,87 @@
...
 
 #define ok_ShellExecuteEx (winetest_set_location(__FILE__, __LINE__), 0) ? (void)0 : TestShellExecuteEx
+#define ok_ShellExecuteEx1 (winetest_set_location(__FILE__, __LINE__), 0) ? (void)0 : TestShellExecuteEx1
+
+
+/* returns TRUE on success and FALSE on failure */
+/* Returns the OS Vendor String in VendorString if found */
+BOOL GetOSVendor(char* VendorString, int length)
+{
+    CHAR file_name[25] = {"OperatingSystem.txt"};
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: braces not required.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 8]
```diff
@@ -2,12 +2,87 @@
...
+    CHAR cmdline[100] = {"ver > "};
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: Ditto.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 9]
```diff
@@ -2,12 +2,87 @@
...
+    fp = fopen(file_name, "r"); // read mode
+
+    if (fp == NULL)
+        return FALSE;
+
+    /* first we must read past the <CR><LF> at the beginning of the file */
+    myreturn = fgets(OSVendor , 100 , fp);
+
+    if (myreturn == NULL)
+        return FALSE;
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: Close fp.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 10]
```diff
@@ -2,12 +2,87 @@
...
+
+    /* Now we can read the first actual text line */
+    myreturn = fgets(OSVendor , 100 , fp);
+
+    if (myreturn == NULL)
+        return FALSE;
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: Close fp. You can use `goto`.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 11]
```diff
@@ -2,12 +2,87 @@
...
+
+    space_ptr = strchr(OSVendor, ' ');
+    pos = -1;
+    if (space_ptr != NULL)
+        {
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: Coding style wrong.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 12]
```diff
@@ -2,12 +2,87 @@
...
+
+    system(cmdline);
+
+    fp = fopen(file_name, "r"); // read mode
+
+    if (fp == NULL)
+        return FALSE;
+
+    /* first we must read past the <CR><LF> at the beginning of the file */
+    myreturn = fgets(OSVendor , 100 , fp);
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: extra space before commas. use `ARRAYSIZE` macro.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 13]
```diff
@@ -2,12 +2,87 @@
...
+
+    if (myreturn == NULL)
+        return FALSE;
+
+    /* Now we can read the first actual text line */
+    myreturn = fgets(OSVendor , 100 , fp);
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: Ditto.

File: modules/rostests/apitests/shell32/ShellExecuteEx.cpp [hunk 14]
```diff
@@ -2,12 +2,87 @@
...
+
+    space_ptr = strchr(OSVendor, ' ');
+    pos = -1;
+    if (space_ptr != NULL)
+        {
+            pos = space_ptr - OSVendor;
+        }
+
+    /* move zero into previous first space character position */
+    if (OSVendor[pos] == 32) OSVendor[pos]=0;
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ShellExecuteEx.cpp]: care the case of `pos == -1`

**Comment by katahiromz** [comment]
We can simply check the `GetWindowsDirectoryW` to detect the OS vendor.

**Pull request closed without merging**